readthedocs/storage/security_token_service.py

```diff
@@ -0,0 +1,154 @@
+     if session_id is given, the name of the session would be "rtd-{session_id}-{project}-{version}".
+     AWS limits the session name to 64 characters, so if the session_id is too long, it will be truncated.
+     For example, for a token used in a build, a good session_id is the ID of the build.
+    :duration: The duration of the credentials in seconds. Default is 15 minutes.
+     Note that the minimum duration time is 15 minutes and the maximum is given by the role (defaults to 1 hour).
+
+    .. note::
+
+       If USING_AWS is set to False, this function will return
+       the values of the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY settings.
+       Useful for local development where we use minio.
+    """
+    if not settings.USING_AWS:
+        return AWSTemporaryCredentials(
+            access_key_id=settings.AWS_ACCESS_KEY_ID,
+            secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
+            # A session token is not needed for the default credentials.
+            session_token=None,
+        )
+
+    bucket_arn = f"arn:aws:s3:::{settings.S3_MEDIA_STORAGE_BUCKET}"
+    storage_paths = version.get_storage_paths()
+    # Generate the list of allowed prefix resources
+    # The resulting prefix looks like:
+    # - html/project/latest/*
+    # - pdf/project/latest/*
+    allowed_prefixes = [f"{storage_path}/*" for storage_path in storage_paths]
+
+    # Generate the list of allowed object resources in ARN format.
+    # The resulting ARN looks like:
+    # arn:aws:s3:::readthedocs-media/html/project/latest/*
+    # arn:aws:s3:::readthedocs-media/pdf/project/latest/*
+    allowed_objects_arn = [f"{bucket_arn}/{prefix}" for prefix in allowed_prefixes]
+
+    # Define an inline policy document to limit the permissions of the temporary credentials.
+    policy_document = {
+        "Version": "2012-10-17",
+        "Statement": [
+            {
+                "Effect": "Allow",
+                "Action": [
+                    "s3:GetObject",
+                    "s3:PutObject",
+                    "s3:DeleteObject",
+                ],
+                "Resource": allowed_objects_arn,
+            },
+            # In order to list the objects in a path, we need to allow the ListBucket action.
+            # But since that action is not scoped to a path, we need to limit it using a condition.
+            {
+                "Effect": "Allow",
+                "Action": ["s3:ListBucket"],
+                "Resource": [
+                    bucket_arn,
+                ],
+                "Condition": {
+                    "StringLike": {
+                        "s3:prefix": allowed_prefixes,
+                    }
+                },
+            },
+        ],
+    }
+
+    session_prefix = f"rtd-{session_id}" if session_id else "rtd"
+    role_session_name = f"{session_prefix}-{project.slug}-{version.slug}"
+    # Limit to 64 characters, as per AWS limitations.
+    role_session_name = role_session_name[:64]
```

Comment:
I assume these don't need to be unique? If so, I could see them colliding if there isn't a session-id

Reply:
yeah, this is meant to identify the "user" we are generating the tokens on behalf of.